core/frontend.ml

```diff
@@ -32,231 +40,275 @@ let show_post_frontend_ast
-    Debug.print ("Pre-Frontend AST:\n" ^ Sugartypes.show_sentence sentence);
-
-  let pre_transformers =
-    interactive_pre_typing_transformers pos_context prev_tyenv in
-  let apply_pre_tc_transformer sentence transformer =
-    transformer sentence in
-  let pre_tc_sentence =
-    List.fold_left
-      apply_pre_tc_transformer
-      sentence
-      pre_transformers in
-
-  let (post_tc_sentence, t, post_tyenv) =
-    TypeSugar.Check.sentence prev_tyenv pre_tc_sentence in
-
-  let apply_post_tc_transformer sentence (name, transformer) =
-    let post = transformer prev_tyenv sentence in
-    if type_check_transformer name then
-      let _ =
-        try TypeSugar.Check.sentence { prev_tyenv with Types.desugared = true } post
-        with e -> trace_type_check_error name Sugartypes.pp_sentence sentence post e
-      in post
-    else
-      post
+
+let trace_type_error transform_name print program program' stacktrace exn =
+  Debug.f "error: transformation pass '%s' produced an ill-typed program.\n%!" transform_name;
+  let print x =
+    let buffer = Buffer.create 1024 in
+    let formatter = Format.formatter_of_buffer buffer in
+    Format.pp_set_margin formatter 120;
+    print formatter x;
+    Format.pp_print_flush formatter ();
+    Buffer.contents buffer
   in
-  let result_sentence =
-    List.fold_left
-      apply_post_tc_transformer
-      post_tc_sentence
-      interactive_post_typing_transformers in
+  Debug.if_set check_frontend_transformations_dump
+    (fun () ->
+      Printf.sprintf "Program before transformation:\n%s\n\nProgram after transformation:\n%s\n\n"
+        (print program) (print program'));
+  Printexc.raise_with_backtrace exn stacktrace
+
+
+(* Pipeline infrastructure. *)
+type 'a result =
+  { program: 'a;
+    datatype: Types.datatype;
+    context: Context.t }
+
+open Transform
+
+(* The [Untyped] module contains the infrastructure for untyped
+   transform passes. *)
+module Untyped = struct
+  type transformer = (module Untyped.S)
+
+  (* This functor constructs a conditional transformer from a given
+     transformer and a (side-effecting) condition. *)
+  module Conditional(T : sig
+               include Untyped.S
+               val condition : unit -> bool
+             end) = struct
+
+    module Untyped = struct
+      let name = Printf.sprintf "Conditional(%s)" T.Untyped.name
+
+      let program state program =
+        if T.condition ()
+        then T.Untyped.program state program
+        else Identity.Untyped.program state program
+
+      let sentence state sentence =
+        if T.condition ()
+        then T.Untyped.sentence state sentence
+        else Identity.Untyped.sentence state sentence
+    end
+  end
+
+  let only_if : bool Settings.setting -> (module Untyped.S) -> transformer
+    = fun setting (module T) ->
+    (module Conditional(struct include T let condition () = Settings.get setting end))
+
+  module Collect_FFI_Files : Untyped.S = struct
```

Comment:
I know that some of this logic was present in `Frontend` before, but maybe this should go somewhere else now

Reply:
Perhaps. I suggest we keep it there for now. In the future we might want to make the compilation pipeline more dynamic, allowing transforms to be plugged in.